File: cases/taint/kotlin/recursive-exception-persistence-negative/RecursiveExceptionPersistenceNegative.kt
package dataflowbench

object RecursiveExceptionPersistenceNegative {
    class Box(var value: Int = 0)

    private class RecursiveSignal : Exception()

    fun dfb_source(): Int { // DFB-SOURCE: recursive-exception-persistence-input
        return 7
    }

    fun dfb_sink(value: Int) {} // DFB-SINK: recursive-exception-persistence-sink

    fun walk(box: Box, value: Int, depth: Int) {
        if (depth == 0) {
            box.value = value // DFB-WITNESS: recursive-exception-persistence-base
            box.value = 0 // DFB-KILL: recursive-exception-persistence-base-clean
            throw RecursiveSignal() // DFB-WITNESS: recursive-exception-persistence-throw
        }
        walk(box, value, depth - 1) // DFB-WITNESS: recursive-exception-persistence-transfer
    }

    fun run() {
        val box = Box()
        try {
            walk(box, dfb_source(), 3)
        } catch (caught: RecursiveSignal) { // DFB-WITNESS: recursive-exception-persistence-catch
            dfb_sink(box.value + 1) // DFB-WITNESS: recursive-exception-persistence-compose
        }
    }
}

File: cases/taint/csharp/recursive-exception-persistence-negative/RecursiveExceptionPersistenceNegative.cs
using System;

namespace DataFlowBench;

static class RecursiveExceptionPersistenceNegative
{
    sealed class Box
    {
        public int Value;
    }

    sealed class RecursiveSignal : Exception
    {
    }

    static int dfb_source() // DFB-SOURCE: recursive-exception-persistence-input
    {
        return 7;
    }

    static void dfb_sink(int value) { } // DFB-SINK: recursive-exception-persistence-sink

    static void Walk(Box box, int value, int depth)
    {
        if (depth == 0)
        {
            box.Value = value; // DFB-WITNESS: recursive-exception-persistence-base
            box.Value = 0; // DFB-KILL: recursive-exception-persistence-base-clean
            throw new RecursiveSignal(); // DFB-WITNESS: recursive-exception-persistence-throw
        }

        Walk(box, value, depth - 1); // DFB-WITNESS: recursive-exception-persistence-transfer
    }

    static void Run()
    {
        Box box = new Box();
        try
        {
            Walk(box, dfb_source(), 3);
        }
        catch (RecursiveSignal) // DFB-WITNESS: recursive-exception-persistence-catch
        {
            dfb_sink(box.Value + 1); // DFB-WITNESS: recursive-exception-persistence-compose
        }
    }
}
